### DISCUSSION TITLE
Introduction of Options and Result natively in .NET

### DISCUSSION THREAD

**Discussion by julienasp** [vkhorikov/CSharpFunctionalExtensions#557] [Ideas]
With the TypeUnions [proposals ](https://github.com/dotnet/csharplang/blob/main/proposals/TypeUnions.md#common-unions)Option type and Result type will be built-in

see a quick [demo ](https://www.youtube.com/watch?v=aksjZkCbIWA)

@vkhorikov have you already think about how you wish to see the lib go? 

we could add support for all the existing extensions methods to support that built-in type? 

like making all Maybe extensions available for Option (.NET type) 
and make all Result extensions available for Result (.NET type)

or we could simply translate an option to a maybe, and a result(.NET) to a result here

anyway it's just food for thoughts

**Comment by vkhorikov** [comment]
Yeah, not sure what route we should take here. I'll need to play around with the built-in Option and Result to understand this better.

The second approach (adding an implicit conversion b/w them) sounds better to me thus far because union types don't behave exactly the same as the current implementations of Result and Maybe. E.g I don't think there will be a check like this https://github.com/vkhorikov/CSharpFunctionalExtensions/blob/master/CSharpFunctionalExtensions/Result/ResultTE.cs#L17
